rfcs/enforce-legibility.md

```diff
@@ -0,0 +1,247 @@
+
+## Examples
+
+The current above configuration contains rules for **removing** (`remove`), **narrowing** (`structure`), **localizing** (`local`) or **moving** (`move`) open statements. They are applied by running the following commands:
+
+```bash
+ocamlclose lint -p fixes file1.ml file2.ml # Prints a set of recommendations, saved in a 'fixes' file
+ocamlclose patch -ic fixes                 # Modifies the original file in place, checking they can still build
+```
+
+Here are some use cases for each rule.
+
+### Remove
+
+Some open statements are just not used enough to justify their existence. For example, the file `src/lib/work_selector/work_lib.ml` contains an `open Currency` which is just used once. `ocaml-close` yields the following patch:
+
+```diff
+ open Core_kernel
+-open Currency
+ open Async
+ 
+ module Make (Inputs : Intf.Inputs_intf) = struct
+@@ -123,7 +122,7 @@
+       (Inputs.Snark_pool.get_completed_work snark_pool statements)
+       ~f:(fun priced_proof ->
+         let competing_fee = Inputs.Transaction_snark_work.fee priced_proof in
+-        Fee.compare fee competing_fee < 0)
++        Currency.Fee.compare fee competing_fee < 0)
+ 
+   module For_tests = struct
+     let does_not_have_better_fee = does_not_have_better_fee
+```
+
+### Structure
+
+When a global statement is used a lot but only for a referring to a few different symbols from the opened module, we can make this set of symbols explicit by using a structured open. For example, in file `src/lib/pickles_types/hlist.ml`, we obtain:
+
+```diff
+ open Core_kernel
+-open Poly_types
++open struct
++  open Poly_types
++  module type T4 = T4
++  module type T3 = T3
++  module type T2 = T2
++  module type T0 = T0
++  module type T1 = T1
++end
++
+ 
+ module E13 (T : T1) = struct
+   type ('a, _, _) t = 'a T.t
+```
+
+Note that currently, the `structure` rule does not match if one of the symbols is a type, even though it would be very useful (often a module is opened only for its single type `t`). This is because opening a type alias does not export the constructures or fields of the original type. Hence in the structured open, we would have to explicitly copy the whole definition of the original type, which would arguably defeat the legibility purpose. Another possibility is the usage of `ppx_import` or `ppx_open`. However this would add new ppxs in the Mina codebase (staged ppxs even, which slow down the compilation since they need access to type information), which would again go against legibility.
+
+### Local
+
+Some open statements are heavily used, but their usage is concentrated in only a few functions. In this case, we can remove the global open and add a few local opens in these functions instead. For example in `src/lib/transition_frontier/persistent_frontier/persistent_frontier.ml`, we can transform two different opens:
+
+```diff
+ open Async_kernel
+ open Core
+ open Mina_base
+-open Mina_state
+ open Mina_transition
+-open Frontier_base
```

Comment:
`Frontier_base` is part of the 'layering' for `Persistent_frontier`.. The `Mina_state` module may make sense to inline, so this example is still okay once that's handled.

Reply:
Done. This uncovered a bug in the "layering" detection, thanks for that!